staging/src/k8s.io/client-go/scale/util.go

```
@@ -0,0 +1,156 @@
+	ScaleForResource(resource schema.GroupVersionResource) (scaleVersion schema.GroupVersionKind, err error)
+}
+
+// discoveryScaleResolver is a ScaleKindResolver that uses
+// a DiscoveryInterface to associate resources with their
+// scale-kinds
+type discoveryScaleResolver struct {
+	discoveryClient discovery.ServerResourcesInterface
+
+	subresMap map[schema.GroupVersionResource]schema.GroupVersion
+	mu        sync.RWMutex
+}
+
+func (r *discoveryScaleResolver) generateKindMap() error {
+	resourceLists, err := r.discoveryClient.ServerResources()
+	if err != nil {
+		return fmt.Errorf("unable to update scale kinds: %v")
+	}
+
+	subresMap := make(map[schema.GroupVersionResource]schema.GroupVersion)
+
+	// first, find the scale subresources that we care about
+	for _, resourceList := range resourceLists {
+		groupVer, err := schema.ParseGroupVersion(resourceList.GroupVersion)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("unable to update scale kinds for API group %s: %v", groupVer.String(), err))
+			continue
+		}
+
+		for _, resource := range resourceList.APIResources {
+			resourceParts := strings.SplitN(resource.Name, "/", 2)
+			if resource.Kind != "Scale" || len(resourceParts) != 2 || resourceParts[1] != "scale" {
+				// skip non-scale resources
+				continue
+			}
+
+			mainResource := groupVer.WithResource(resourceParts[0])
+			scaleGV := groupVer
+			if resource.Group != "" && resource.Version != "" {
+				scaleGV = schema.GroupVersion{
+					Group:   resource.Group,
+					Version: resource.Version,
+				}
+			}
+
+			subresMap[mainResource] = scaleGV
+		}
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.subresMap = subresMap
+
+	return nil
+}
+
+func (r *discoveryScaleResolver) ScaleForResource(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
+	r.mu.RLock()
+	gv, exists := r.subresMap[resource]
+	if !exists {
+		// retry on misses in case we have an update
+		r.mu.RUnlock()
+		r.generateKindMap()
```

ignoring an error doesn't look right.